### DISCUSSION TITLE
Understanding IPvAnyAdress using pyright

### DISCUSSION THREAD

**Discussion by phaabe** [pydantic/pydantic#10312] [Question] [answered]
This example [from the documentation](https://docs.pydantic.dev/latest/api/networks/#pydantic.networks.IPvAnyAddress) is not correctly validated by pyright:

```py
from pydantic import BaseModel
from pydantic.networks import IPvAnyAddress

class IpModel(BaseModel):
    ip: IPvAnyAddress

print(IpModel(ip='127.0.0.1'))
```

Pyright's error:
```"Literal['127.0.0.1']" cannot be assigned to parameter "ip" of type "IPvAnyAddress"```

I tried fixing this by explicitly typing:
```py
print(IpModel(ip=IPvAnyAddress('127.0.0.1')))
```

This does not work wither because `IPvAnyAddress('127.0.0.1')` returns the type `IPv4Address`.

**Fix 1**
obviously
`# type:ignore`

**Fix 2**
use `IPv4Address` from `ipaddress` 
```py
from pydantic import BaseModel
from ipaddress import IPv4Address

class IpModel(BaseModel):
    ip: IPv4Address

print(IpModel(ip=IPv4Address('127.0.0.1')))
```

What are your thoughts on the problem and the two solutions? Can someone elaborate on the reason of the original problem?

**Comment by Viicos** [comment, marked as answer]
This is a more general limitation with [`dataclass_transform`](https://docs.python.org/3/library/typing.html#typing.dataclass_transform), which cannot take into account the type conversion. This example will also fail:

```python
from pydantic import BaseModel

class Foo(BaseModel):
    bar: int

Foo(bar="1")
```

You can make use of the [`mypy` plugin](https://docs.pydantic.dev/dev/integrations/mypy/), however I'm not sure if it covers the `Json` type yet, to be tested.
